### ISSUE TITLE
[Bug] nameserver-policy按照域名指定DNS失效

### ISSUE THREAD

**Issue by dycwuxing** [MetaCubeX/mihomo#882]
### Verify steps

- [X]  确保你使用的是**本仓库**最新的的 mihomo 或 mihomo Alpha 版本 Ensure you are using the latest version of Mihomo or Mihomo Alpha from **this repository**. 
- [X]  如果你可以自己 debug 并解决的话，提交 PR 吧 Is this something you can **debug and fix**? Send a pull request! Bug fixes and documentation fixes are welcome. 
- [X]  我已经在 [Issue Tracker](……/) 中找过我要提出的问题 I have searched on the [issue tracker](……/) for a related issue. 
- [X]  我已经使用 Alpha 分支版本测试过，问题依旧存在 I have tested using the dev branch, and the issue still exists. 
- [X]  我已经仔细看过 [Documentation](https://wiki.metacubex.one/) 并无法自行解决问题 I have read the [documentation](https://wiki.metacubex.one/) and was unable to solve the issue. 
- [X]  这是 Mihomo 核心的问题，并非我所使用的 Mihomo 衍生版本（如 OpenMihomo、KoolMihomo 等）的特定问题 This is an issue of the Mihomo core *per se*, not to the derivatives of Mihomo, like OpenMihomo or KoolMihomo. 

### Mihomo version

alpha-c5d1db7(created at Thu Dec 7 08:00:21 CST 2023)

### What OS are you seeing the problem on?

Windows

### Mihomo config

```yaml
dns:
  enable: true
  ipv6: true
  ipv6-timeout: 300
  enhanced-mode: redir-host
  use-hosts: true
  listen: 0.0.0.0:7874
  max-ttl: 60
  default-nameserver:
  - tls://223.5.5.5:853
  - https://1.12.12.12/dns-query
  - https://223.5.5.5/dns-query
  nameserver:
  - https://1.1.1.1/dns-query
  - https://1.0.0.1/dns-query
  - tls://1.1.1.1
  proxy-server-nameserver:
  - https://doh.pub/dns-query
  - https://dns.alidns.com/dns-query
  - https://doh.360.cn
  nameserver-policy:
    www.baidu.com: 114.114.114.114
    rule-set:ChinaMax_Classical:
    - tls://223.5.5.5:853
    - https://1.12.12.12/dns-query
    - https://223.5.5.5/dns-query
```

### Mihomo log

```shell
10615	[info]	[TCP] 127.0.0.1:6095(msedge.exe) --> www.baidu.com:443 match GeoIP(cn) using 🎯全球直连[DIRECT]
10614	[debug]	[DNS] cache hit for www.baidu.com., expire at 2023-12-07 09:05:50
10613	[debug]	[DNS] cache hit for www.baidu.com., expire at 2023-12-07 09:05:24
10612	[debug]	[DNS] www.baidu.com --> 220.181.38.150
10611	[debug]	[DNS] www.baidu.com --> [240e:83:205:5a:0:ff:b05f:346b 240e:83:205:58:0:ff:b09f:36bf] AAAA from https://223.5.5.5:443/dns-query
10610	[debug]	[DNS] www.baidu.com --> [220.181.38.150 220.181.38.149] A from https://223.5.5.5:443/dns-query
10609	[debug]	[DNS] resolve www.baidu.com from https://1.12.12.12:443/dns-query
10608	[debug]	[DNS] resolve www.baidu.com from tls://223.5.5.5:853
10607	[debug]	[DNS] resolve www.baidu.com from tls://223.5.5.5:853
10606	[debug]	[DNS] resolve www.baidu.com from https://223.5.5.5:443/dns-query
10605	[debug]	[DNS] resolve www.baidu.com from https://1.12.12.12:443/dns-query
10604	[debug]	[DNS] resolve www.baidu.com from https://223.5.5.5:443/dns-query
```

### Description

因为公司内网域名需要指定DNS才能正确解析,今日最新的alpha无法打开公司域名
经用baidu测试,发现指定域名并未使用指定DNS进行解析导致
(配置未变的情况下,前期核心无此问题,V1.17也无此问题,疑似我之前提报ruleset进行DNS分流失效的fix之后发生)

**Comment by dycwuxing** [comment]
> ![image](https://private-user-images.githubusercontent.com/78135608/288785261-72c7793c-1e38-465e-8bff-86fd4703e17d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE5OTY0NjIsIm5iZiI6MTcwMTk5NjE2MiwicGF0aCI6Ii83ODEzNTYwOC8yODg3ODUyNjEtNzJjNzc5M2MtMWUzOC00NjVlLThiZmYtODZmZDQ3MDNlMTdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA4VDAwNDI0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVmM2QwMDgzMDYzZWVlNzk3NWMyZGExZWRjYzlmYzQ1ZjQ0ZDA2ZTAxNWM0NDdlMTNhMjRmODJmNjJmYWEzN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.v_dQuQJeWa3-9rtwzGWGyA3lh0IUYtq3W5yco2BSoxo) 无法复现

大佬,请再帮助看一下,我用的今天的"alpha-9ac4738.zip"还是一样
下附我单独做的配置文件,我这边还是能复现
[li1wgXbYUl6q.zip](https://github.com/MetaCubeX/mihomo/files/13607666/li1wgXbYUl6q.zip)


![image](https://github.com/MetaCubeX/mihomo/assets/44988562/277dae01-1149-4daa-88e6-bda956a18209)
![企业微信截图_17019960402136](https://github.com/MetaCubeX/mihomo/assets/44988562/7fb3eec8-4a23-4e6d-8553-67bf0489477b)

**Comment by wwqgtxx** [comment]
fixed in: https://github.com/MetaCubeX/mihomo/commit/1d1841f7aad3f45526aec0000511c8ea3b64e9fe

**Comment by dycwuxing** [comment]
> fixed in: [1d1841f](https://github.com/MetaCubeX/mihomo/commit/1d1841f7aad3f45526aec0000511c8ea3b64e9fe)

大佬太效率了!!!
alpha-1d1841f已正常(Fri Dec 8 09:04:27 CST 2023)
![image](https://github.com/MetaCubeX/mihomo/assets/44988562/9d8a50cb-d0c4-4725-ba4f-efef90db62cb)
